python/nx-cugraph/nx_cugraph/classes/graph.py

```diff
@@ -40,57 +45,236 @@
+                return Gcg._cudagraph
+            return Gcg
+        if self.__dict__["_node"] is None:
+            raise RuntimeError(
+                f"{type(self).__name__} cannot be converted to the GPU, because it is "
+                "not on the CPU! This is not supposed to be possible. If you believe "
+                "you have found a bug, please report a minimum reproducible example to "
+                "https://github.com/rapidsai/cugraph/issues/new/choose"
+            )
+        try:
+            Gcg = nxcg.from_networkx(
+                self, preserve_edge_attrs=True, preserve_node_attrs=True
+            )
+        except Exception:
+            # Should we warn that the full graph can't be on GPU?
+            nx_cache[_CANT_CONVERT_TO_GPU] = True
+            return None
+        Gcg.graph = self.graph
+        cache[_CACHE_KEY] = Gcg
+        return Gcg
+
+    @_cudagraph.setter
+    def _cudagraph(self, val, *, clear_cpu=True):
+        """Set the full ``CudaGraph`` for this graph, or remove from device if None."""
+        if (cache := getattr(self, "__networkx_cache__", None)) is None:
+            # Should we warn?
+            return
+        # TODO: pay close attention to when we should clear the cache, since
+        # this may or may not be a mutation.
+        cache = cache.setdefault("backends", {}).setdefault("cugraph", {})
+        if val is None:
+            cache.pop(_CACHE_KEY, None)
+        else:
+            self.graph = val.graph
+            cache[_CACHE_KEY] = val
+            if clear_cpu:
+                for key in self._nx_attrs:
+                    self.__dict__[key] = None
+
+    @nx.Graph.name.setter
+    def name(self, s):
+        # Don't clear the cache when setting the name, since `.graph` is shared.
+        # There is a very small risk here for the cache to become (slightly)
+        # insconsistent if graphs from other backends are cached.
+        self.graph["name"] = s
+
+    @classmethod
+    @networkx_api
+    def is_directed(cls) -> bool:
+        return False
+
+    @classmethod
+    @networkx_api
+    def is_multigraph(cls) -> bool:
+        return False
+
+    @classmethod
+    def to_cudagraph_class(cls) -> type[CudaGraph]:
+        return CudaGraph
+
+    @classmethod
+    @networkx_api
+    def to_directed_class(cls) -> type[nxcg.DiGraph]:
+        return nxcg.DiGraph
+
+    @classmethod
+    def to_networkx_class(cls) -> type[nx.Graph]:
+        return nx.Graph
+
+    @classmethod
+    @networkx_api
+    def to_undirected_class(cls) -> type[Graph]:
+        return Graph
+
+    def __init__(self, incoming_graph_data=None, **attr):
+        super().__init__(incoming_graph_data, **attr)
+        self.__networkx_cache__ = _GraphCache(self)
+
+    def _reify_networkx(self) -> None:
+        """Copy graph to host (CPU) if necessary."""
+        if self.__dict__["_node"] is None:
+            # After we make this into an nx graph, we rely on the cache being correct
```

Comment:
If we have graphs in the cache _and_ reified networkx data structured (e.g., `_adj` and `_node`), then we want all the graphs to be consistent. Hence, the graph can be on host and device at the same time--it doesn't move back and forth keeping only one. Consistency requires the cache being "correct" such as being cleared when the networkx data structures are mutated. Notably, the inner dicts of the data structures are just dicts. The cache is cleared when using methods such as `G.add_edge`, but not when doing e.g. `G[u][v][key] = val`. If we wanted to be super-careful and strict, we could make our own mutable mapping class that clears the cache upon mutation, but this would come at a (possibly significant) performance penalty. This is what the cache warning message discusses.

Reply:
Right I understand that, but I was asking more because the comment about the cache just seemed out of place.  After looking at the code again, I remembered `self._cudagraph` is actually a property that uses the cache, which I'm assuming is what the comment is referring to.  Assuming this is correct, I think I just needed a reminder that `self._cudagraph` could end up using a cached graph.